packages/@lwc/synthetic-shadow/src/polyfills/window-event-target/polyfill.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { windowRemoveEventListener, windowAddEventListener } from '../../env/window';
+
+import { defineProperties, isUndefined, isTrue, ArraySlice } from '@lwc/shared';
+import { eventTargetGetter } from '../../env/dom';
+import { isNodeShadowed } from '../../shared/node-ownership';
+
+// this method returns true if an event can be seen by a particular eventTarget,
+// otherwise it returns false, which means the listener will never be invoked.
+function isQualifyingEventTarget(currentTarget: Window, evt: Event): boolean {
+    const { composed } = evt;
+    const originalTarget = eventTargetGetter.call(evt);
+    if (originalTarget === currentTarget) {
+        // dispatched on the window directly
+        return true;
+    }
+    // if the event reaches the window by propagation is because it is bubbling,
+    // in which the the only it really depends on the composed value
+    return isTrue(composed) || !isNodeShadowed(originalTarget as Node);
+}
+
+function getEventListenerWrapper(fnOrObj: EventListenerOrEventListenerObject): EventListener {
+    let wrapperFn: EventListener | undefined = (fnOrObj as any).$$lwcEventWrapper$$;
+    if (isUndefined(wrapperFn)) {
+        wrapperFn = (fnOrObj as any).$$lwcEventWrapper$$ = function(this: Window, e: Event) {
+            if (isQualifyingEventTarget(this, e)) {
+                if (typeof fnOrObj === 'function') {
+                    fnOrObj.call(this, e);
```

Comment:
This snippet is repeated from polyfill/event-target/polyfill.ts, the only difference being the type specification of `this`. We should move this to a common util.

Also, if the same listener is used on window and document(or any extensions of EventTarget), the same cached wrapper will be used on both because the property names(.$$lwcEventWrapper$$) are the same. This is fine today because the wrapper logic is the same in both the polyfills. In the future if we change logic, then we might be accessing the wrong wrapper. I would prefer if we explicitly specify that the wrapper logic is the same by moving `getEventListenerWrapper` to a util or using a different property name to cache the wrapped listener.

Reply:
ok

Reply:
I prefer to keep them separate... for now... it is not a lot of code.

as for the second part, lets just use a weakmap to be safe.